[#38] 조건을 활용해 사용자를 검색할 수 있는 기능 구현하기

msugo1 opened the pull request:
📖 Model 관련

    📑  SearchConditionsRequest

    - 검색조건에 대한 요청을 받아 컨트롤러 계층으로 전달하는 객체를 만들 클래스

    📑 SerachConditions

    - 검증을 통과한 검색 조건들과 다른 필요한 조건들을 모두 모아 DB로 전달하기 전 하나의 객체로 만들어 줄때 사용될 클래스

    📖 Controller 계층 관련

    📑 ProfileController

    - 조건을 설정해서 사용자를 검색하는 요청을 받는 `searchUserProfiles` 메소드 추가

    - validator와 `@Valid` 어노테이션을 활용해 `@RequestBody`를 통해 요청받은 값들을 일차적으로 검증할 수 있도록 설계

   📖 Service 계층 관련

    📑 ProfileService

    - 검색조건에 대한 로직을 처리할 `searchUserProfiles` 메소드 추가

    - 검색조건 중 학습 언어에 대한 언어 레벨 검증 실시(NATIVE 레벨이 조건에 추가되어 있지는 않은지)

    - 검증이 끝나면 요청받은 조건과 현재 로그인 중인 사용자의 아이디 그리고 cursor 기반 페이징 값을 모두 합쳐서 하나의 `conditions` 객체 생성

    - 특히, 나이의 경우 기본 최소, 최댓값을 부여함으로써 null 값이 들어오더라도 DB에서 유연하게 처리할 수 있도록 시도

    📑 LanguageService

    - 언어 레벨 검증에 대한 로직이 `ProfileService`에도 사용이 되므로 기존에 private 메소드로 LanguageService 내부에 존재했던 검증 메소드를 외부 static 메소드로 추출, 별도의 클래스 생성

    - 로직 변경에 따라 언어 레벨만을 담은 리스트를 추출해주는 private 메소드를 대신 생성 후, 해당 리스트를 위의 검증 메소드로 전달할 수 있도록 로직 변경

    📖 Mapper 계층 관련

    📑 ProfileMapper

    - 전달 받은 검색 조건을 가지고 DB에서 적절한 값을 찾아올 `searchUserProfiles` 메소드 추가

    📑 ProfileMapper.xml

    - 위의 검색 조건을 처리할 수 있도록 쿼리 작성 및 튜닝

    Annotation 관련

    📑 AgeRange

    - 커스텀 Validator 적용을 표시할 커스텀 어노테이션 생성

    - validation 위반 발생 시 예외를 던지면서 함께 전달할 기본 메시지 추가

    📖 Util 관련

    📑 AgeRangeValidator

    - 나이를 검색 조건으로 전달 받는 경우 터무니 없는 숫자가 들어오는 것을 막기 위해 나이의 최소값과 최대값을 설정할 필요

    - 기존의 @min, @max 어노테이션을 활용하면 쉽게 정수 값을 검증할 수 있었으나 개별 값들만 검증 가능하고, 연계된 값들을 검증하는데 사용할 수 있는 validator는 없었기에 커스텀 validator를 생성

    - 최솟값은 10살, 최댓값은 99살으로 설정하고 좀 더 유연한 유지보수를 위해 두 값을 상수로 생성해서 사용할 수 있도록 만듦

    - 나이 조건은 nullable 하므로 각 조건을 4개의 분기로 나눠 요청으로 들어오는 조건이 최대, 최소값을 초과할 수 없도록 하였으며, 최솟값의 경우에도 최댓값을 넘을 수 없도록 코드 작성

    📑 LanguageRangeValidator

    - 기존 언어 레벨을 검증하는 메소드는 LanguageService 내부에 private 메소드로 별도로 존재했으나, 프로필을 검색하면서 레벨 조건을 설정하기 때문에 맞는 레벨이 들어왔는지 검증할 필요성이 생김

    - 중복 코드 작성을 없애기 위해서 별도의 Validator 클래스를 생성하고 내부에 레벨 검증 메소드를 static 메소드로 둠으로써 필요한 클래스들에서  바로 활용할 수 있도록 변경

   📖  테스트 코드 작성

    🔖 TestUsersFixture

    - 테스트에 사용할 유저들의 Fixture을 한 곳에 모아놓은 클래스 생성

    🔖 ITSearchProfilesTest

    - 필수 조건과 선택 조건을 적용한 통합 테스트 메소드 작성 후 적절한 결과가 나오는지 검증

    🔖 ProfileService

    - 컨트롤러 요청으로 들어온 언어 레벨에 중복되는 값이 있는지 검사할 필요가 없도록 List 자체를 Set으로 변경

    🔖 LanguageLevelValidator

    - `List와 Set 타입이 모두` 적절한 언어 레벨인지 검증할 수 있는 메소드를 사용할 수 있도록 파라미터의 타입을 List에서 `Collection`으로 변경

    🔖 User

    - 성별을 입력하는 칸에 정규표현식을 추가해서 원하는 값만 들어올 수 있도록 검증과정 추가

Commit:
[#38] 사용자 프로필 검색기능 구현

Model 관련

* SearchConditionsRequest

- 검색조건에 대한 요청을 받아 컨트롤러 계층으로 전달하는 객체를 만들 클래스

* SerachConditions

- 검증을 통과한 검색 조건들과 다른 필요한 조건들을 모두 모아 DB로
  전달하기 전 하나의 객체로 만들어 줄때 사용될 클래스

Controller 계층 관련

* ProfileController

- 조건을 설정해서 사용자를 검색하는 요청을 받는 `searchUserProfiles`
  메소드 추가

- validator와 `@Valid` 어노테이션을 활용해 `@RequestBody`를 통해
  요청받은 값들을 일차적으로 검증할 수 있도록 설계

Service 계층 관련

* ProfileService

- 검색조건에 대한 로직을 처리할 `searchUserProfiles` 메소드 추가

- 검색조건 중 학습 언어에 대한 언어 레벨 검증 실시(NATIVE 레벨이 조건에
  추가되어 있지는 않은지)

- 검증이 끝나면 요청받은 조건과 현재 로그인 중인 사용자의 아이디 그리고
  cursor 기반 페이징 값을 모두 합쳐서 하나의 `conditions` 객체 생성

- 특히, 나이의 경우 기본 최소, 최댓값을 부여함으로써 null 값이
  들어오더라도 DB에서 유연하게 처리할 수 있도록 시도

* LanguageService

- 언어 레벨 검증에 대한 로직이 `ProfileService`에도 사용이 되므로
  기존에 private 메소드로 LanguageService 내부에 존재했던 검증 메소드를
  외부 static 메소드로 추출, 별도의 클래스 생성

- 로직 변경에 따라 언어 레벨만을 담은 리스트를 추출해주는 private 메소드를 대신 생성 후, 해당 리스트를 위의 검증 메소드로 전달할 수 있도록 로직 변경

Mapper 계층 관련

* ProfileMapper

- 전달 받은 검색 조건을 가지고 DB에서 적절한 값을 찾아올
  `searchUserProfiles` 메소드 추가

* ProfileMapper.xml

- 위의 검색 조건을 처리할 수 있도록 쿼리 작성 및 튜닝

Annotation 관련

* AgeRange

- 커스텀 Validator 적용을 표시할 커스텀 어노테이션 생성

- validation 위반 발생 시 예외를 던지면서 함께 전달할 기본 메시지 추가

Util 관련

* AgeRangeValidator

- 나이를 검색 조건으로 전달 받는 경우 터무니 없는 숫자가 들어오는 것을
  막기 위해 나이의 최소값과 최대값을 설정할 필요가 있었음

- 기존의 @min, @max 어노테이션을 활용하면 쉽게 정수 값을 검증할 수 있었으나 개별 값들만 검증 가능하고, 연계된 값들을 검증하는데 사용할 수 있는 validator는 없었기에 커스텀 validator를 생성해야 했음

- 최솟값은 10살, 최댓값은 99살으로 설정하고 좀 더 유연한 유지보수를 위해
  두 값을 상수로 생성해서 사용할 수 있도록 만듦

- 나이 조건은 nullable 하므로 각 조건을 4개의 분기로 나눠 요청으로
  들어오는 조건이 최대, 최소값을 초과할 수 없도록 하였으며, 최솟값의
  경우에도 최댓값을 넘을 수 없도록 코드 작성

* LanguageRangeValidator

- 기존 언어 레벨을 검증하는 메소드는 LanguageService 내부에 private
  메소드로 별도로 존재했으나, 프로필을 검색하면서 레벨 조건을 설정하기
  때문에 맞는 레벨이 들어왔는지 검증할 필요성이 생김

- 중복 코드 작성을 없애기 위해서 별도의 Validator 클래스를 생성하고
  내부에 레벨 검증 메소드를 static 메소드로 둠으로써 필요한 클래스들에서
  바로 활용할 수 있도록 변경

Commit:
Merge branch 'feature/37' into feature/38

Commit:
[#38] 프로필 검색 Validator 테스트 작성

* SearchProfileValidatorTest

- 각 필수/선택 조건에 대한 성공, 실패 범위 설정 후 단위 테스트 작성

* 정적 메소드나 정적 변수만을 가진 클래스들이 의도와는 상관없이
  인스턴스화 되는 것을 막기 위해 `@NoArgsConstructor에 Private 접근
  제어자 부여`

Commit:
[#38] 사용자 검색 조건에 대한 통합 테스트 작성

* TestUsersFixture

- 테스트에 사용할 유저들의 Fixture을 한 곳에 모아놓은 클래스 생성

* ITSearchProfilesTest

- 필수 조건과 선택 조건을 적용한 통합 테스트 메소드 작성 후 원하는
  결과가 나오는지 확인

* ProfileService

- 컨트롤러 요청으로 들어온 언어 레벨에 중복되는 값이 있는지 검사할 필요가 없도록 List 자체를 Set으로 변경

* LanguageLevelValidator

- `List와 Set 타입이 모두` 적절한 언어 레벨인지 검증할 수 있는 메소드를
  사용할 수 있도록 파라미터의 타입을 List에서 `Collection`으로 변경

* User

- 성별을 입력하는 칸에 정규표현식을 추가해서 원하는 값만 들어올 수
  있도록 검증

msugo1 commented on src/main/java/me/soo/helloworld/util/validator/AgeRangeValidator.java:
```diff
@@ -0,0 +1,32 @@
+
+    public static final int MIN_AGE_RANGE = 10;
+
+    @Override
```
커스텀 Validator를 조건에 맞춰서 만들긴 했는데 분기가 많다보니 `if - else`를 지나치게 많이 사용한감이 없지 않아 있는 것 같은데요. 테스트 코드를 만들면서 계속 따로 고민해봤는데 조금 더 깔끔하게 만들 수 있는 방법이 떠오르지가 않아서요. 혹시 어떻게 개선 가능할 지 조언좀 주시면 감사하겠습니다!

f-lab-dev replied:
일단 복잡한 구문들을 메소드로 하나하나 분리하는것부터 시작해보면 어떨까요?

msugo1 replied:
여러 방법들을 시도해 봤는데 잘 안되서 일단은 복잡한 구문들만 메소드 추출로 빼서 정리했습니다. Null Check 하는 if else 구문이 들어가야 하다보니까 생각보다 잘 안되네요;

msugo1 commented on src/main/resources/mappers/ProfileMapper.xml:
```diff
@@ -93,4 +94,52 @@
+        FROM users AS u
+            INNER JOIN speak s ON u.userId = s.userId
+            INNER JOIN speak l ON u.userId = l.userId
+        WHERE
```
쿼리가 조금 복잡해서 인덱스도 다시 걸어가면서, EXPLAIN 확인해서 최대한 튜닝을 하긴 했는데요. 동작은 제가 의도했던 대로 하는 것 같은데요. 쿼리에 조건을 이렇게 많이 걸어본 적이 사실 처음이긴 한데 너무 복잡하게 적혀있는 것 같아서요. 혹시 어떤식으로 고민해보면 더 개선을 해볼수가 있을까요?

f-lab-dev replied:
일단 지금 공부한거로는 최대한 튜닝을 해보신 것 같은데 일단 튜닝을 어떤 과정을 거쳐서 했냐가 중요한 것 같습니다~ 블로그로 한번 정리해보시면 보면서 같이 얘기해봐도 좋을 것 같아요.

그리고 확실히 조건이 엄청 복잡한데 저렇게 조건을 많이 걸게되면 조건에 따라 걸리게 되는 쿼리가 달라지기 때문에 최적화 한게 안먹힐수도 있어서 저 복잡한 조건문을 어떻게 할건지도 고민을 해보시면 좋을 것 같습니다~

Commit:
[#38] 통합 테스트 패키지 이동

- ITSearchProfilesTest를 통합테스트가 모여있는 패키지로 이동

f-lab-dev commented on src/main/java/me/soo/helloworld/model/condition/SearchConditionsRequest.java:
```diff
@@ -0,0 +1,49 @@
+
+@Getter
+@Builder
+@AgeRange
```
개별 데이터마다 검증이 필요한거니까 컬럼에 달아주는게 어떨까요??

msugo1 replied:
한 클래스 내에 있는 minAge, maxAge의 두 변수를 교차 검증하는 방법하려면 두 변수를 한 번에 가져와서 비교를 할 수 있는 방법을 찾아야 하는데 클래스 레벨의 어노테이션 없이 검증하는 방법을 검색으로 찾기가 쉽지가 않네요 😢

f-lab-dev commented on src/main/java/me/soo/helloworld/model/condition/SearchConditionsRequest.java:
```diff
@@ -0,0 +1,49 @@
+    private final Integer livingTown;
+
+    @NotNull(message = "사용자 검색 시 상대방의 구사언어 정보는 필수로 입력하셔야 합니다.")
+    private final Integer speakLanguage;
```
not null이면 primitive type으로 선언하시는게 어떨까요?

msugo1 replied:
처음 설계할 때는 `@Nullable`로 설계했던지라 Integer로 해줬었는데 `@NotNull`로 변환 후 거기까지 바꿔주는 것을 깜빡했네요. 수정하도록 하겠습니다.

f-lab-dev commented on src/main/java/me/soo/helloworld/service/LanguageService.java:
```diff
@@ -70,7 +72,9 @@ public void modifyLanguageLevels(String userId, List<LanguageData> languageNewLe
         }
 
-        validateLevel(languageNewLevels, status);
+        List<LanguageLevel> newLangLevels = extractLanguageLevelsOnly(languageNewLevels);
+        LanguageLevelValidator.validateLevel(newLangLevels, status);
```
이 두줄이 계속 중복되는것같은데 `extractLanguageLevelsOnly`에서 validation까지 해주는건 어떨까요?

msugo1 replied:
코드 중복을 생각을 못하고 있었네요.

대신, 리팩토링을 한다면 `extractLanguageLevelsOnly 의 존재 이유`가 `validateLevel을 위해 List<LanguageData>에서 언어 레벨만을 따로 추출`해주는 것이기 반대로 validateLevel으로 언어 정보에 대한 데이터를 넘겨주고 다시 이곳에서 레벨을 뽑아내도록 리팩토링 하는 것은 어떨지 궁금합니다.

`LanguageService`는 예전부터 다른 곳도 조금 리팩토링 해야겠다고 생각하고 있어서 따로 브랜치를 하나 파서 작업해도 괜찮을까요?

Commit:
[#38] 피드백 반영

* AgeRangeValidaotr 코드 리팩토링

- 여러 분기를 메소드 추출을 활용해 리팩토링 (차후 더 리팩토링 예정)

* SearchConditionsRequest

- `@NotNull` 어노테이션이 달린 Integer 타입의 변수를 int로 변경

- int 최솟값이 1부터 시작할 수 있도록 `@Min(1)` 부여

- 변경에 따른 테스트 코드 수정

msugo1 committed:
Merge branch 'develop' into feature/38
